File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BaseConditionalSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.ExecutionException;

import eu.pretix.libpretixsync.api.ApiException;
import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.api.ResourceNotModified;
import eu.pretix.libpretixsync.db.RemoteObject;
import eu.pretix.libpretixsync.db.ResourceSyncStatus;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;

abstract public class BaseConditionalSyncAdapter<T extends RemoteObject & Persistable, K> extends BaseDownloadSyncAdapter<T, K> {
    public BaseConditionalSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, eventSlug, api, feedback);
    }

    private PretixApi.ApiResponse firstResponse;

    @Override
    protected JSONObject downloadPage(String url, boolean isFirstPage) throws ApiException, ResourceNotModified {
        ResourceSyncStatus resourceSyncStatus = store.select(ResourceSyncStatus.class)
                .where(ResourceSyncStatus.RESOURCE.eq(getResourceName()))
                .and(ResourceSyncStatus.EVENT_SLUG.eq(eventSlug))
                .limit(1)
                .get().firstOrNull();
        if (resourceSyncStatus == null) {
            resourceSyncStatus = new ResourceSyncStatus();
        } else {
            if (!getMeta().equals(resourceSyncStatus.getMeta()) && !(getMeta().equals("") && resourceSyncStatus.getMeta() == null)) {
                store.delete(resourceSyncStatus);
                resourceSyncStatus = new ResourceSyncStatus();
            }
        }
        PretixApi.ApiResponse apiResponse = api.fetchResource(url, resourceSyncStatus.getLast_modified());
        if (isFirstPage) {
            firstResponse = apiResponse;
        }
        return apiResponse.getData();
    }

    public String getMeta() {
        return "";
    }

    @Override
    public void download() throws JSONException, ApiException, ExecutionException, InterruptedException {
        firstResponse = null;
        super.download();
        if (firstResponse != null) {
            ResourceSyncStatus resourceSyncStatus = store.select(ResourceSyncStatus.class)
                    .where(ResourceSyncStatus.RESOURCE.eq(getResourceName()))
                    .and(ResourceSyncStatus.EVENT_SLUG.eq(eventSlug))
                    .limit(1)
                    .get().firstOrNull();
            if (resourceSyncStatus == null) {
                resourceSyncStatus = new ResourceSyncStatus();
                resourceSyncStatus.setResource(getResourceName());
                resourceSyncStatus.setEvent_slug(eventSlug);
                resourceSyncStatus.setMeta(getMeta());
            }
            if (firstResponse.getResponse().header("Last-Modified") != null) {
                resourceSyncStatus.setLast_modified(firstResponse.getResponse().header("Last-Modified"));
                resourceSyncStatus.setMeta(getMeta());
                store.upsert(resourceSyncStatus);
            }
            firstResponse = null;
        }
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BaseConditionalSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import eu.pretix.libpretixsync.api.ApiException
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.api.ResourceNotModified
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import org.json.JSONException
import org.json.JSONObject
import java.util.Date
import java.util.concurrent.ExecutionException

abstract class BaseConditionalSyncAdapter<T, K>(
    db: SyncDatabase,
    api: PretixApi,
    syncCycleId: String,
    eventSlug: String,
    fileStorage: FileStorage,
    feedback: ProgressFeedback?,
) : BaseDownloadSyncAdapter<T, K>(
    db = db,
    api = api,
    syncCycleId = syncCycleId,
    eventSlug = eventSlug,
    fileStorage = fileStorage,
    feedback = feedback,
) {
    private var firstResponse: PretixApi.ApiResponse? = null

    @Throws(ApiException::class, ResourceNotModified::class)
    override fun downloadPage(url: String, isFirstPage: Boolean): JSONObject? {
        val resourceSyncStatus = db.resourceSyncStatusQueries.selectByResourceAndEventSlug(
            resource = getResourceName(),
            event_slug = eventSlug,
        ).executeAsOneOrNull()

        val lastModified = if (resourceSyncStatus == null) {
            null
        } else {
            if (getMeta() != resourceSyncStatus.meta && !(getMeta() == "" && resourceSyncStatus.meta == null)) {
                db.resourceSyncStatusQueries.deleteById(resourceSyncStatus.id)
                Date().toString()
            } else {
                resourceSyncStatus.last_modified
            }
        }
        val apiResponse = api.fetchResource(url, lastModified)

        if (isFirstPage) {
            firstResponse = apiResponse
        }
        return apiResponse.data
    }

    protected open fun getMeta(): String {
        return ""
    }

    @Throws(
        JSONException::class,
        ApiException::class,
        ExecutionException::class,
        InterruptedException::class
    )
    override fun download() {
        firstResponse = null
        super.download()

        val currentResponse = firstResponse
        if (currentResponse != null) {
            val resourceSyncStatus = db.resourceSyncStatusQueries.selectByResourceAndEventSlug(
                resource = getResourceName(),
                event_slug = eventSlug,
            ).executeAsOneOrNull()

            val lastModified = if (currentResponse.response.header("Last-Modified") != null) {
                currentResponse.response.header("Last-Modified")
            } else {
                null
            }

            if (lastModified != null && resourceSyncStatus == null) {
                db.resourceSyncStatusQueries.insert(
                    event_slug = eventSlug,
                    last_modified = lastModified,
                    resource = getResourceName(),
                    meta = getMeta(),
                    status = null,
                )
            } else if (lastModified != null && resourceSyncStatus != null) {
                db.resourceSyncStatusQueries.updateLastModifiedAndMeta(
                    last_modified = lastModified,
                    meta = getMeta(),
                    id = resourceSyncStatus.id,
                )
            }
            firstResponse = null
        }
    }
}
